programs/util.c

```diff
@@ -640,10 +640,42 @@ int UTIL_countPhysicalCores(void)
     }
 }
 
-#elif defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__) || defined(__DragonFly__)
+#elif defined(__FreeBSD__)
 
-/* Use apple-provided syscall
- * see: man 3 sysctl */
+#include <sys/param.h>
+#include <sys/sysctl.h>
+
+/* Use physical core sysctl when available
+ * see: man 4 smp, man 3 sysctl */
+int UTIL_countPhysicalCores(void)
+{
+    static int numPhysicalCores = 0; /* freebsd sysctl is native int sized */
+    if (numPhysicalCores != 0) return numPhysicalCores;
+
+#if __FreeBSD_version >= 1300008
+    {   size_t size = sizeof(numPhysicalCores);
+        int ret = sysctlbyname("kern.smp.cores", &numPhysicalCores, &size, NULL, 0);
+        if (ret == 0) return numPhysicalCores;
+        if (errno != ENOENT) {
+            perror("zstd: can't get number of physical cpus");
+            exit(1);
```

Comment:
Could it be recoverable instead of exit?

Reply:
It is possible to add, but I don't think a recovery path makes sense here.

I followed the existing convention from the `__APPLE__` [section above](https://github.com/facebook/zstd/pull/1491/files#diff-0d55920a6e9a92f8408dbf7a76c994bfR567) for unexpected sysctl failures.

If the `sysctl` succeeds, we're done; if the sysctl fails, and the `errno` was `ENOENT` (i.e., we're running on an older kernel predating the interface), we have a recovery path via the older `sysconf` method.

The only case where this termination occurs is when `sysctl` returns error status and `errno` is not `ENOENT`.  The only plausible scenarios for that occurring are programmer error (userspace side), which does not seem to be the case; or a radical change in kernel sysctl topology (`kern.smp.cores` is modified into an intermediary node with leaves below it, e.g., `kern.smp.cores.foo`).  I think those scenarios can be set aside, for much the same reason(s) they are set aside on MacOS.